### ISSUE TITLE
std::chrono::year_month_day support

### ISSUE THREAD

**Issue by Alfrink** [jtv/libpqxx#667]
Hi,

according to the description of the 7.7.0 release, support for `std::chrono::year_month_day` should be available, even if "pretty limited" in some way. However, I'm having trouble using that type at all:
```
work.exec_params(
            "INSERT INTO [...] VALUES ($1 [...]) ",
            someStruct.yearMonthDayVariable [...]
    );
```
yields `error: the value of ‘pqxx::nullness<std::chrono::year_month_day, void>::always_null’ is not usable in a constant expression` during compilation. Also:
```
work.exec_params(
            "INSERT INTO [...] VALUES ($1 [...]) ",
            pqxx::to_string( someStruct.yearMonthDayVariable ) [...]
    );
```
yields `undefined reference to `pqxx::nullness<std::chrono::year_month_day, void>::is_null(std::chrono::year_month_day const&)'` during linking.

What am I doing wrong?

Thank you!

**Comment by Alfrink** [comment]
My project's C++ standard is set properly in my top level CMakeLists.txt:
```
set(CMAKE_CXX_STANDARD 20)
[...]
FetchContent_Declare(pqxx
        GIT_REPOSITORY https://github.com/jtv/libpqxx
        GIT_TAG 17e5a6c) # 7.7.4
```
If I was using C++17, I would get `year_month_day is not a member of std::chrono` errors, as it only got added in C++20.

**Comment by tt4g** [comment]
If the compiler supports it, `PQXX_HAVE_YEAR_MONTH_DAY` should be defined in `include/pqxx/config-internal-compiler.h`. Is it defined?

https://github.com/jtv/libpqxx/blob/99f9658977c86999d2137a421583afedd5ab128e/cmake/config.cmake#L110-L113

https://github.com/jtv/libpqxx/blob/99f9658977c86999d2137a421583afedd5ab128e/cmake/config.cmake#L143-L157

https://github.com/jtv/libpqxx/blob/99f9658977c86999d2137a421583afedd5ab128e/include/pqxx/config.h.in#L111-L112

**Comment by Alfrink** [comment]
Yes, it's defined. Code in an `#ifdef PQXX_HAVE_YEAR_MONTH_DAY` block gets executed. I'm on gcc 11.2.

**Comment by jtv** [comment]
That means that the build, when you ran cmake, did not detect support for this type.

I _still_ haven't figured out where cmake logs these checks.  But if you tried the `configure` script (and setting `-std=c++20` in your `CXXFLAGS` while doing so), it should write the build output to `config.log`.

**Comment by tt4g** [comment]
Could you define `PQXX_HAVE_YEAR_MONTH_DAY` macro yourself to resolve the compile error? If the compile error is resolved, then you have not imported the header file correctly.

**Comment by Alfrink** [comment]
Manually doing `#define PQXX_HAVE_YEAR_MONTH_DAY` doesn't seem to help.

Running `./configure CXXFLAGS='-std=c++20'` manually, `config.log` contains:
```
configure:17688: checking for std::chrono::year_month_day etc
configure:17709: g++ -o conftest -std=c++20 -fvisibility=hidden -fvisibility-inlines-hidden   conftest.cpp  >&5
configure:17709: $? = 0
configure:17719: result: yes
```
The CMake counterpart would be `CMakeFiles/CMakeConfigureLog.yaml`, which was just added to CMake 3.26. I just updated to that version for this purpose, and after executing cmake for my project (which declares its pqxx dependency using CMake's FetchContent module), it contains:
```
kind: "try_compile-v1"
    backtrace:
      - "cmake-build-debug/_deps/pqxx-src/cmake/config.cmake:113 (try_compile)"
      - "cmake-build-debug/_deps/pqxx-src/CMakeLists.txt:27 (include)"
    directories:
      source: "[...]/cmake-build-debug/_deps/pqxx-build/CMakeFiles/CMakeTmp"
      binary: "[...]/cmake-build-debug/_deps/pqxx-build/CMakeFiles/CMakeTmp"
    cmakeVariables:
      CMAKE_CXX_FLAGS: "--coverage -Wall -Wextra"
      CMAKE_CXX_FLAGS_DEBUG: "-g -g"
      CMAKE_EXE_LINKER_FLAGS: ""
      CMAKE_MODULE_PATH: "[...]/cmake-build-debug/_deps/pqxx-src/cmake"
    buildResult:
      variable: "PQXX_HAVE_YEAR_MONTH_DAY"
      cached: true
      stdout: |
        Change Dir: [...]/cmake-build-debug/_deps/pqxx-build/CMakeFiles/CMakeTmp
        
        Run Build Command(s):/usr/bin/cmake -E env VERBOSE=1 /usr/bin/gmake -f Makefile cmTC_9c9c0/fast && /usr/bin/gmake  -f CMakeFiles/cmTC_9c9c0.dir/build.make CMakeFiles/cmTC_9c9c0.dir/build
        gmake[1]: Entering directory '[...]/cmake-build-debug/_deps/pqxx-build/CMakeFiles/CMakeTmp'
        Building CXX object CMakeFiles/cmTC_9c9c0.dir/year_month_day.cxx.o
        /usr/bin/c++   --coverage -Wall -Wextra  -std=c++20 -fdiagnostics-color=always -o CMakeFiles/cmTC_9c9c0.dir/year_month_day.cxx.o -c [...]/cmake-build-debug/_deps/pqxx-src/config-tests/year_month_day.cxx
        Linking CXX executable cmTC_9c9c0
        /usr/bin/cmake -E cmake_link_script CMakeFiles/cmTC_9c9c0.dir/link.txt --verbose=1
        /usr/bin/c++ --coverage -Wall -Wextra  CMakeFiles/cmTC_9c9c0.dir/year_month_day.cxx.o -o cmTC_9c9c0 
        gmake[1]: Leaving directory '[...]/cmake-build-debug/_deps/pqxx-build/CMakeFiles/CMakeTmp'
        
      exitCode: 0
```
, which looks fine to me. And as expected, my project's CMakeCache.txt contains:
```
//Result of TRY_COMPILE
PQXX_HAVE_YEAR_MONTH_DAY:INTERNAL=TRUE
```

**Comment by tt4g** [comment]
> Manually doing `#define PQXX_HAVE_YEAR_MONTH_DAY` doesn't seem to help.

Maybe you have libpqxx installed on your machine and the header files from that are included?
If the installed header file has `#undef PQXX_HAVE_YEAR_MONTH_DAY`, then `#define PQXX_HAVE_YEAR_MONTH_DAY` may not make sense.

**Comment by jtv** [comment]
Oh argh.  I'm sorry - I missed an important detail.  Bit busy, trying to follow along on the go using my phone.

That `config.h.in` file wasn't the right file to look at.  That's the _template_ for the config header.  We should've been looking at `config.h`.  And based on what followed, I expect it did define the macro all along.

But...  could you try `#include <pqxx/time>` for me?  It's possible that I just forgot to add it to the `<pqxx/pqxx>` header.

**Comment by Alfrink** [comment]
Including `<pqxx/time>` fixed it, I completely missed that the conversions have their own include... especially silly since it's the same with ranges and I'm already including `<pqxx/range>`. I guess the initial error message just didn't look like a missing include to me at all. Anyway, thanks for your help!
